zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+    block: Arc<Block>,
+    /// `block`'s cached header hash.
+    hash: BlockHeaderHash,
+    /// The transmitting end of a oneshot channel.
+    ///
+    /// The receiving end of this oneshot is passed to the caller as the future.
+    /// It has two layers of `Result`s, an inner `checkpoint::Error`, and an
+    /// outer `tokio::sync::oneshot::error::RecvError`.
+    tx: oneshot::Sender<Result<BlockHeaderHash, Error>>,
+}
+
+/// A list of unverified blocks at a particular height.
+///
+/// Typically contains zero or one blocks, but might contain more if a peer
+/// has an old chain fork. (Or sends us a bad block.)
+type QueuedBlockList = Vec<QueuedBlock>;
 
+/// A block height verification range.
+///
+/// Implements `RangeBounds<BlockHeight>`.
+type VerifyBounds = (Bound<BlockHeight>, Bound<BlockHeight>);
+
+/// A checkpointing block verifier.
+///
+/// Verifies blocks using a supplied list of checkpoints. There must be at
+/// least one checkpoint for the genesis block.
+#[derive(Debug)]
+struct CheckpointVerifier {
+    // Inputs
+    //
     /// Each checkpoint consists of a coinbase height and block header hash.
     ///
     /// Checkpoints should be chosen to avoid forks or chain reorganizations,
     /// which only happen in the last few hundred blocks in the chain.
     /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
     /// orphaned side-chains after 288 blocks.)
-    checkpoint_list: Arc<HashMap<BlockHeight, BlockHeaderHash>>,
+    ///
+    /// There must be a checkpoint for the genesis block at BlockHeight 0.
+    /// (All other checkpoints are optional.)
+    checkpoint_list: BTreeMap<BlockHeight, BlockHeaderHash>,
+
+    // Queued Blocks
+    //
+    /// A queue of unverified blocks.
+    ///
+    /// Contains a list of unverified blocks at each block height. In most cases,
+    /// the checkpoint verifier will store zero or one block at each height.
+    ///
+    /// Blocks are verified in order, when there is a chain from the next
+    /// checkpoint, back to the `current_checkpoint_height`.Each pair of
+    /// checkpoints is used to verify all the blocks between those checkpoints.
+    ///
+    /// The first checkpoint does not have any ancestors, so it only verifies the
+    /// genesis block.
+    queued: BTreeMap<BlockHeight, QueuedBlockList>,
+
+    /// The range of heights that we are currently verifying. Extends from the
+    /// most recently verified checkpoint (`Excluded`), to the next highest
+    /// checkpoint (`Included`).
+    ///
+    /// If checkpoint verification has not started yet, the current range only
+    /// contains the genesis checkpoint.
+    ///
+    /// `None` means that checkpoint verification has finished.
+    current_checkpoint_range: Option<VerifyBounds>,
 }
 
-/// The error type for the CheckpointVerifier Service.
-// TODO(jlusby): Error = Report ?
-type Error = Box<dyn error::Error + Send + Sync + 'static>;
+/// The CheckpointVerifier implementation.
```

Comment:
TIL you can associate docs with impl blocks _and_ they show up in rustdoc before all the fns in that impl block